<?xml version="1.0" ?>
<robot name="robot">
  <link name="world_link">
    <inertial>
      <mass value="0"/>
      <inertia ixx="0.2125" ixy="-0.005" ixz="0.0225" iyy="0.205" iyz="0.045" izz="0.0125"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </inertial>
  </link>
  <link name="x_prismatic_link">
    <inertial>
      <mass value="0.01"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </inertial>
  </link>
  <joint name="x_to_world_joint" type="prismatic">
    <parent link="world_link"/>
    <child link="x_prismatic_link"/>
    <axis xyz="1 0 0"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <limit effort="0.0" velocity="1000.0"/>
  </joint>
  <link name="y_prismatic_link">
    <inertial>
      <mass value="0.01"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </inertial>
  </link>
  <joint name="y_to_x_joint" type="prismatic">
    <parent link="x_prismatic_link"/>
    <child link="y_prismatic_link"/>
    <axis xyz="0 1 0"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <limit effort="0.0" velocity="1000.0"/>
  </joint>
  <!-- Note the z-offset to raise the base off the ground, to avoid collision issues. Without this, there are vibrations and other artifacts from resolving the ground collision that result in significant error in the balanced object. -->
  <joint name="base_to_y_joint" type="continuous">
    <parent link="y_prismatic_link"/>
    <child link="base_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0.01"/>
  </joint>
  <material name="dark_grey">
    <color rgba="0.2 0.2 0.2 1.0"/>
  </material>
  <material name="light_grey">
    <color rgba="0.4 0.4 0.4 1.0"/>
  </material>
  <material name="red">
    <color rgba="0.8 0.0 0.0 1.0"/>
  </material>
  <material name="white">
    <color rgba="0.9 0.9 0.9 1.0"/>
  </material>
  <material name="yellow">
    <color rgba="0.8 0.8 0.0 1.0"/>
  </material>
  <material name="black">
    <color rgba="0.15 0.15 0.15 1.0"/>
  </material>
  <material name="seethrough">
    <color rgba="1.0 0.0 0.0 0.25"/>
  </material>
  <link name="base_link">
    <inertial>
      <mass value="0.01"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </inertial>
  </link>
  <!-- Transform between base and arm can be calibrated and is loaded from parameters. -->
  <joint name="base_link_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="base_link"/>
    <child link="chassis_link"/>
  </joint>
  <link name="chassis_link">
    <visual>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/ridgeback/body.stl"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/ridgeback/body-collision.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="1.57079632679 0 1.57079632679" xyz="0.012 0.002 0.067"/>
      <!-- <origin rpy="1.57079632679 0 1.57079632679" xyz="0 0 0"/> -->
      <mass value="165.304"/>
      <inertia ixx="4.4744" ixy="0.03098" ixz="0.003647" iyy="7.1624" iyz="0.1228" izz="4.6155"/>
    </inertial>
  </link>
  <joint name="right_side_cover_link_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="chassis_link"/>
    <child link="right_side_cover_link"/>
  </joint>
  <joint name="left_side_cover_link_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="chassis_link"/>
    <child link="left_side_cover_link"/>
  </joint>
  <link name="left_side_cover_link">
    <visual>
      <origin rpy="0 0 3.14159265359" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/ridgeback/side-cover.stl"/>
      </geometry>
      <material name="yellow"/>
    </visual>
  </link>
  <link name="right_side_cover_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/ridgeback/side-cover.stl"/>
      </geometry>
      <material name="yellow"/>
    </visual>
  </link>
  <joint name="front_cover_link_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="chassis_link"/>
    <child link="front_cover_link"/>
  </joint>
  <link name="front_cover_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/ridgeback/end-cover.stl"/>
      </geometry>
      <material name="black"/>
    </visual>
  </link>
  <joint name="rear_cover_link_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="chassis_link"/>
    <child link="rear_cover_link"/>
  </joint>
  <link name="rear_cover_link">
    <visual>
      <origin rpy="0 0 3.14159265359" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/ridgeback/end-cover.stl"/>
      </geometry>
      <material name="black"/>
    </visual>
  </link>
  <joint name="front_lights_link_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="chassis_link"/>
    <child link="front_lights_link"/>
  </joint>
  <joint name="rear_lights_link_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="chassis_link"/>
    <child link="rear_lights_link"/>
  </joint>
  <link name="front_lights_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/ridgeback/lights.stl"/>
      </geometry>
      <material name="white"/>
    </visual>
  </link>
  <link name="rear_lights_link">
    <visual>
      <origin rpy="0 0 3.14159265359" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/ridgeback/lights.stl"/>
      </geometry>
      <material name="red"/>
    </visual>
  </link>
  <joint name="top_link_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="chassis_link"/>
    <child link="top_link"/>
  </joint>
  <link name="top_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/ridgeback/top.stl"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/ridgeback/top.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="arch_link">
    <visual>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/mobile_manipulation_central/meshes/tor11_mech.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
  </link>
  <joint name="arch_joint" type="fixed">
    <origin rpy="0 0 1.57079632679" xyz="0 0 0.278"/>
    <!-- <origin rpy="-1.57079632679 0 1.57079632679" xyz="0 0 0.278"/> -->
    <parent link="top_link"/>
    <child link="arch_link"/>
  </joint>
  <!--
    Base UR robot series xacro macro.

    NOTE: this is NOT a URDF. It cannot directly be loaded by consumers
    expecting a flattened '.urdf' file. See the top-level '.xacro' for that
    (but note: that .xacro must still be processed by the xacro command).

    For use in '.launch' files: use one of the 'load_urX.launch' convenience
    launch files.

    This file models the base kinematic chain of a UR robot, which then gets
    parameterised by various configuration files to convert it into a UR3(e),
    UR5(e), UR10(e) or UR16e.

    NOTE: the default kinematic parameters (ie: link lengths, frame locations,
    offets, etc) do not correspond to any particular robot. They are defaults
    only. There WILL be non-zero offsets between the Forward Kinematics results
    in TF (ie: robot_state_publisher) and the values reported by the Teach
    Pendant.

    For accurate (and robot-specific) transforms, the 'kinematics_parameters_file'
    parameter MUST point to a .yaml file containing the appropriate values for
    the targetted robot.

    If using the UniversalRobots/Universal_Robots_ROS_Driver, follow the steps
    described in the readme of that repository to extract the kinematic
    calibration from the controller and generate the required .yaml file.

    Main author of the migration to yaml configs: Ludovic Delval.

    Contributors to previous versions (in no particular order):

     - Felix Messmer
     - Kelsey Hawkins
     - Wim Meeussen
     - Shaun Edwards
     - Nadia Hammoudeh Garcia
     - Dave Hershberger
     - G. vd. Hoorn
     - Philip Long
     - Dave Coleman
     - Miguel Prada
     - Mathias Luedtke
     - Marcel Schnirring
     - Felix von Drigalski
     - Felix Exner
     - Jimmy Da Silva
     - Ajit Krisshna N L
     - Muhammad Asif Rana
  -->
  <!--
    NOTE: the macro defined in this file is NOT part of the public API of this
          package. Users CANNOT rely on this file being available, or stored in
          this location. Nor can they rely on the existence of the macro.
  -->
  <transmission name="ur10_arm_shoulder_pan_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="ur10_arm_shoulder_pan_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="ur10_arm_shoulder_pan_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="ur10_arm_shoulder_lift_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="ur10_arm_shoulder_lift_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="ur10_arm_shoulder_lift_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="ur10_arm_elbow_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="ur10_arm_elbow_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="ur10_arm_elbow_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="ur10_arm_wrist_1_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="ur10_arm_wrist_1_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="ur10_arm_wrist_1_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="ur10_arm_wrist_2_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="ur10_arm_wrist_2_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="ur10_arm_wrist_2_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="ur10_arm_wrist_3_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="ur10_arm_wrist_3_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="ur10_arm_wrist_3_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!-- links: main serial chain -->
  <link name="ur10_arm_base_link"/>
  <link name="ur10_arm_base_link_inertia">
    <visual>
      <origin rpy="0 0 3.141592653589793" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/visual/base.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 3.141592653589793" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/collision/base.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="4.0"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.0061063308908" ixy="0.0" ixz="0.0" iyy="0.0061063308908" iyz="0.0" izz="0.01125"/>
    </inertial>
  </link>
  <link name="ur10_arm_shoulder_link">
    <visual>
      <origin rpy="0 0 3.141592653589793" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/visual/shoulder.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 3.141592653589793" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/collision/shoulder.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="7.778"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.03147431257693659" ixy="0.0" ixz="0.0" iyy="0.03147431257693659" iyz="0.0" izz="0.021875624999999996"/>
    </inertial>
  </link>
  <link name="ur10_arm_upper_arm_link">
    <visual>
      <origin rpy="1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0.220941"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/visual/upperarm.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0.220941"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/collision/upperarm.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="12.93"/>
      <origin rpy="0 1.5707963267948966 0" xyz="-0.306 0.0 0.175"/>
      <inertia ixx="0.42175380379841093" ixy="0.0" ixz="0.0" iyy="0.42175380379841093" iyz="0.0" izz="0.03636562499999999"/>
    </inertial>
  </link>
  <link name="ur10_arm_forearm_link">
    <visual>
      <origin rpy="1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0.049042"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/visual/forearm.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0.049042"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/collision/forearm.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="3.87"/>
      <origin rpy="0 1.5707963267948966 0" xyz="-0.28615 0.0 0.049042"/>
      <inertia ixx="0.11106969409710458" ixy="0.0" ixz="0.0" iyy="0.11106969409710458" iyz="0.0" izz="0.010884375"/>
    </inertial>
  </link>
  <link name="ur10_arm_wrist_1_link">
    <visual>
      <!-- TODO: Move this to a parameter -->
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 -0.1149"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/visual/wrist1.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 -0.1149"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/collision/wrist1.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.96"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.005108247956699999" ixy="0.0" ixz="0.0" iyy="0.005108247956699999" iyz="0.0" izz="0.005512499999999999"/>
    </inertial>
  </link>
  <link name="ur10_arm_wrist_2_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 -0.1158"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/visual/wrist2.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 -0.1158"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/collision/wrist2.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.96"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.0040792483683" ixy="0.0" ixz="0.0" iyy="0.0040792483683" iyz="0.0" izz="0.005512499999999999"/>
    </inertial>
  </link>
  <link name="ur10_arm_wrist_3_link">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 -0.0922"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/visual/wrist3.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 -0.0922"/>
      <geometry>
        <mesh filename="file:///home/adam/phd/code/projects/upright/catkin_ws/src/tps/universal_robot/ur_description/meshes/ur10/collision/wrist3.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.202"/>
      <origin rpy="0 0 0" xyz="0.0 0.0 -0.01525"/>
      <inertia ixx="0.00011792166116465" ixy="0.0" ixz="0.0" iyy="0.00011792166116465" iyz="0.0" izz="0.00020452500000000002"/>
    </inertial>
  </link>
  <!-- joints: main serial chain -->
  <joint name="ur10_arm_base_link-base_link_inertia" type="fixed">
    <parent link="ur10_arm_base_link"/>
    <child link="ur10_arm_base_link_inertia"/>
    <!-- 'base_link' is REP-103 aligned (so X+ forward), while the internal
           frames of the robot/controller have X+ pointing backwards.
           Use the joint between 'base_link' and 'base_link_inertia' (a dummy
           link/frame) to introduce the necessary rotation over Z (of pi rad).
      -->
    <origin rpy="0 0 3.141592653589793" xyz="0 0 0"/>
  </joint>
  <joint name="ur10_arm_shoulder_pan_joint" type="revolute">
    <parent link="ur10_arm_base_link_inertia"/>
    <child link="ur10_arm_shoulder_link"/>
    <origin rpy="0 0 -6.730070847457299e-05" xyz="0 0 0.1282560065126062"/>
    <axis xyz="0 0 1"/>
    <limit effort="330.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="2.0943951023931953"/>
    <dynamics damping="0" friction="0"/>
  </joint>
  <joint name="ur10_arm_shoulder_lift_joint" type="revolute">
    <parent link="ur10_arm_shoulder_link"/>
    <child link="ur10_arm_upper_arm_link"/>
    <origin rpy="1.570741458413116 0 2.08117439685017e-05" xyz="-0.0001298567256802955 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="330.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="2.0943951023931953"/>
    <dynamics damping="0" friction="0"/>
  </joint>
  <joint name="ur10_arm_elbow_joint" type="revolute">
    <parent link="ur10_arm_upper_arm_link"/>
    <child link="ur10_arm_forearm_link"/>
    <origin rpy="3.13961720923506 3.141368906580845 -3.141538943105826" xyz="-0.6127577422067012 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="150.0" lower="-3.141592653589793" upper="3.141592653589793" velocity="3.141592653589793"/>
    <dynamics damping="0" friction="0"/>
  </joint>
  <joint name="ur10_arm_wrist_1_joint" type="revolute">
    <parent link="ur10_arm_forearm_link"/>
    <child link="ur10_arm_wrist_1_link"/>
    <origin rpy="3.138117507051141 -3.140562763851374 3.141583822174732" xyz="-0.5716744596847277 0.0005679491125303003 0.1634310438150313"/>
    <axis xyz="0 0 1"/>
    <limit effort="56.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.141592653589793"/>
    <dynamics damping="0" friction="0"/>
  </joint>
  <joint name="ur10_arm_wrist_2_joint" type="revolute">
    <parent link="ur10_arm_wrist_1_link"/>
    <child link="ur10_arm_wrist_2_link"/>
    <origin rpy="1.570713528713072 0 -5.018411682050918e-05" xyz="-1.744730553720786e-05 -0.1156796703530661 9.578054833275211e-06"/>
    <axis xyz="0 0 1"/>
    <limit effort="56.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.141592653589793"/>
    <dynamics damping="0" friction="0"/>
  </joint>
  <joint name="ur10_arm_wrist_3_joint" type="revolute">
    <parent link="ur10_arm_wrist_2_link"/>
    <child link="ur10_arm_wrist_3_link"/>
    <origin rpy="1.571028722462297 3.141592653589793 -3.141565446353047" xyz="7.947493484628657e-05 0.09205088147541438 2.139222642043671e-05"/>
    <axis xyz="0 0 1"/>
    <limit effort="56.0" lower="-6.283185307179586" upper="6.283185307179586" velocity="3.141592653589793"/>
    <dynamics damping="0" friction="0"/>
  </joint>
  <!-- ROS-Industrial 'base' frame: base_link to UR 'Base' Coordinates transform -->
  <link name="ur10_arm_base"/>
  <joint name="ur10_arm_base_link-base_fixed_joint" type="fixed">
    <!-- Note the rotation over Z of pi radians: as base_link is REP-103
           aligned (ie: has X+ forward, Y+ left and Z+ up), this is needed
           to correctly align 'base' with the 'Base' coordinate system of
           the UR controller.
      -->
    <origin rpy="0 0 3.141592653589793" xyz="0 0 0"/>
    <parent link="ur10_arm_base_link"/>
    <child link="ur10_arm_base"/>
  </joint>
  <!-- ROS-Industrial 'flange' frame: attachment point for EEF models -->
  <link name="ur10_arm_flange"/>
  <joint name="ur10_arm_wrist_3-flange" type="fixed">
    <parent link="ur10_arm_wrist_3_link"/>
    <child link="ur10_arm_flange"/>
    <origin rpy="0 -1.5707963267948966 -1.5707963267948966" xyz="0 0 0"/>
  </joint>
  <!-- ROS-Industrial 'tool0' frame: all-zeros tool frame -->
  <link name="ur10_arm_tool0"/>
  <joint name="ur10_arm_flange-tool0" type="fixed">
    <!-- default toolframe: X+ left, Y+ up, Z+ front -->
    <origin rpy="1.5707963267948966 0 1.5707963267948966" xyz="0 0 0"/>
    <parent link="ur10_arm_flange"/>
    <child link="ur10_arm_tool0"/>
  </joint>
  <!-- Join arm to the base -->
  <joint name="ur10_arm_base_joint" type="fixed">
    <parent link="arch_link"/>
    <child link="ur10_arm_base_link"/>
    <origin rpy="0.0 0.0 3.14159265359" xyz="0.01 -0.27 0.375"/>
  </joint>
  <!-- Frame coincident with all-zeros TCP on UR controller -->
  <link name="gripper">
    <visual>
      <!-- start halfway along the cylinder -->
      <origin rpy="0 0 0" xyz="0 0 0.145"/>
      <geometry>
        <cylinder length="0.290" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.145"/>
      <geometry>
        <cylinder length="0.290" radius="0.01"/>
      </geometry>
    </collision>
  </link>
  <joint name="tool_gripper_joint" type="fixed">
    <!-- The 15-degree offset about the z-axis is to account for the fact that the gripper is actually mounted at this offset. -->
    <origin rpy="0 0 0.2617993877991494" xyz="0 0.0 0"/>
    <parent link="ur10_arm_tool0"/>
    <child link="gripper"/>
  </joint>
  <link name="ft_sensor"/>
  <joint name="tool_ft_sensor_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.02"/>
    <parent link="ur10_arm_tool0"/>
    <child link="ft_sensor"/>
  </joint>
  <!-- Frame for a gripped object or tool. Its pose is specified in the `transform_params` config file.  -->
  <!-- Something like this should be added in your own project's URDF and customized to your needs. -->
  <!-- <link name="gripped_object" /> -->
  <!-- <joint name="gripped_object_joint" type="fixed"> -->
  <!--   <xacro:transformed_origin name="gripped_object_transform" /> -->
  <!--   <parent link="gripper"/> -->
  <!--   <child link="gripped_object"/> -->
  <!-- </joint> -->
</robot>
